cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -506,22 +506,49 @@ Result<ExecNode*> Declaration::AddToPlan(ExecPlan* plan,
   return node;
 }
 
+Declaration* Declaration::Root() {
+  Declaration* current = this;
+  while (!current->inputs.empty()) {
+    DCHECK_LE(current->inputs.size(), 1)
+        << "No clear root when a declaration has multiple inputs";
```

Comment:
DCHECKing here seems overly aggressive since we don't have anything like `bool has_root()`. Instead could we have
- Root() returns nullptr when no root can be found
- Roots() returns a vector of all decls with no inputs

Reply:
This function has gone away.